browser/ui/views/location_bar/onion_location_view.cc

```diff
@@ -5,176 +5,100 @@
-  }
-
-  OnionLocationButtonView(const OnionLocationButtonView&) = delete;
-  OnionLocationButtonView& operator=(const OnionLocationButtonView&) = delete;
+}  // namespace
 
-  ~OnionLocationButtonView() override = default;
+OnionLocationView::OnionLocationView(
+    Profile* profile,
+    IconLabelBubbleView::Delegate* icon_label_bubble_delegate,
+    PageActionIconView::Delegate* page_action_icon_delegate)
+    : PageActionIconView(/*command_updater=*/nullptr,
+                         /*command_id=*/0,
+                         icon_label_bubble_delegate,
+                         page_action_icon_delegate,
+                         "Tor",
+                         /*ephemeral*/ false),
+      profile_(profile) {
+  SetVisible(false);
+}
 
-  void SetOnionLocation(const GURL& location) {
-    onion_location_ = location;
-    SetTooltipText(l10n_util::GetStringFUTF16(
-        profile_->IsTor() ? IDS_LOCATION_BAR_ONION_AVAILABLE_TOOLTIP_TEXT
-                          : IDS_LOCATION_BAR_OPEN_IN_TOR_TOOLTIP_TEXT,
-        base::UTF8ToUTF16(onion_location_.spec())));
-  }
+OnionLocationView::~OnionLocationView() = default;
 
- private:
-  // views::View
-  void Layout(PassKey) override {
-    LayoutSuperclass<views::LabelButton>(this);
-    UpdateBorder();
-  }
+const gfx::VectorIcon& OnionLocationView::GetVectorIcon() const {
+  return kLeoProductTorIcon;
+}
 
-  void UpdateBorder() {
-    SetBackground(
-        views::CreateRoundedRectBackground(kOpenInTorBg, kCornerRadius));
+void OnionLocationView::UpdateImpl() {
+  if (!GetWebContents()) {
+    SetVisible(false);
+    return;
   }
+  auto* helper = tor::OnionLocationTabHelper::FromWebContents(GetWebContents());
+  const bool show_icon = helper && helper->should_show_icon();
 
-  void ButtonPressed() {
-    TorProfileManager::SwitchToTorProfile(profile_, onion_location_);
+  if (show_icon) {
+    if (profile_->IsTor()) {
+      SetIconColor(kOnionButtonTextColor);
+      SetTextColor(views::Button::STATE_DISABLED, kOnionButtonTextColor);
+      SetEnabledTextColors(kOnionButtonTextColor);
+      SetBackground(views::CreateRoundedRectBackground(
+          kOnionButtonBackgound, kOnionButtonCornerRadius));
+
+      label()->SetVisible(true);
+      SetLabel(brave_l10n::GetLocalizedResourceUTF16String(
+                   IDS_LOCATION_BAR_ONION_AVAILABLE),
+               l10n_util::GetStringFUTF16(
+                   IDS_LOCATION_BAR_ONION_AVAILABLE_TOOLTIP_TEXT,
+                   base::UTF8ToUTF16(helper->onion_location().spec())));
+    } else {
+      if (const ui::ColorProvider* color_provider = GetColorProvider()) {
+        const SkColor icon_color_default =
+            color_provider->GetColor(kColorOmniboxResultsIcon);
+        SetIconColor(icon_color_default);
+      }
+      SetLabel({}, l10n_util::GetStringFUTF16(
+                       IDS_LOCATION_BAR_OPEN_IN_TOR_TOOLTIP_TEXT,
+                       base::UTF8ToUTF16(helper->onion_location().spec())));
```

Comment:
nit: `base::UTF8ToUTF16(helper->onion_location().spec())` should be put in `show_icon` block before `if/else`